tests/xdc/history_replication_conflict_test.go

```diff
@@ -0,0 +1,391 @@
+	}
+}
+
+func (s *historyReplicationConflictTestSuite) SetupTest() {
+	s.tv = testvars.New(s.T().Name())
+	s.setupTest()
+}
+
+func (s *historyReplicationConflictTestSuite) TearDownSuite() {
+	s.tearDownSuite()
+}
+
+// TestConflictResolutionReappliesSignals creates a split-brain scenario in which both clusters believe they are active.
+// Both clusters then accept a signal and write it to their own history, and the test confirms that the signal is
+// reapplied during the resulting conflict resolution process.
+func (s *historyReplicationConflictTestSuite) TestConflictResolutionReappliesSignals() {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, testTimeout)
+	defer cancel()
+	sdkClient1, sdkClient2 := s.createSdkClients()
+
+	runId := s.startWorkflowAndEnterSplitBrainState(ctx, sdkClient1)
+
+	// Both clusters now believe they are active and hence both will accept a signal.
+
+	// Send signals
+	s.NoError(sdkClient1.SignalWorkflow(ctx, s.tv.WorkflowID(), runId, "my-signal", "cluster1-signal"))
+	s.NoError(sdkClient2.SignalWorkflow(ctx, s.tv.WorkflowID(), runId, "my-signal", "cluster2-signal"))
+
+	// cluster1 has accepted a signal
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowExecutionSignaled {"Input": {"Payloads": [{"Data": "\"cluster1-signal\""}]}}
+	`, s.getHistory(ctx, s.cluster1, runId), []int{1, 1, 1})
+
+	// cluster2 has also accepted a signal (with failover version 2 since it is endogenous to cluster 2)
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowExecutionSignaled {"Input": {"Payloads": [{"Data": "\"cluster2-signal\""}]}}
+	`, s.getHistory(ctx, s.cluster2, runId), []int{1, 1, 2})
+
+	// Execute pending history replication tasks. Both clusters believe they are active, therefore each cluster sends
+	// its signal to the other, triggering conflict resolution.
+	s.executeHistoryReplicationTasksUntil(ctx, "cluster1", enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_SIGNALED)
+	s.executeHistoryReplicationTasksUntil(ctx, "cluster2", enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_SIGNALED)
+
+	// cluster1 has received a signal with failover version 2 which superseded its own signal.
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowExecutionSignaled {"Input": {"Payloads": [{"Data": "\"cluster2-signal\""}]}}
+	`, s.getHistory(ctx, s.cluster1, runId), []int{1, 1, 2})
+
+	// cluster2 has reapplied the signal from cluster 1 on top of its own signal.
+	s.HistoryRequire.EqualHistoryEventsAndVersions(`
+	1 WorkflowExecutionStarted
+	2 WorkflowTaskScheduled
+	3 WorkflowExecutionSignaled {"Input": {"Payloads": [{"Data": "\"cluster2-signal\""}]}}
+	4 WorkflowExecutionSignaled {"Input": {"Payloads": [{"Data": "\"cluster1-signal\""}]}}
+	`, s.getHistory(ctx, s.cluster2, runId), []int{1, 1, 2, 2})
+}
```

Comment:
nit: maybe add a comment explaining that history in both clusters will eventually coverage if more replication tasks are executed. I worry people reading the test will get the wrong impression that history will diverge in split brain case.